python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -281,17 +281,181 @@ def partition_time_window(self) -> TimeWindow:
         return self._step_execution_context.partition_time_window
 
+    @public
+    def has_tag(self, key: str) -> bool:
+        """Check if a logging tag is set.
+
+        Args:
+            key (str): The tag to check.
+
+        Returns:
+            bool: Whether the tag is set.
+        """
+        return self._step_execution_context.has_tag(key)
+
+    @public
+    def get_tag(self, key: str) -> Optional[str]:
+        """Get a logging tag.
+
+        Args:
+            key (tag): The tag to get.
+
+        Returns:
+            Optional[str]: The value of the tag, if present.
+        """
+        return self._step_execution_context.get_tag(key)
+
+    def has_events(self) -> bool:
+        return bool(self._events)
+
+    def consume_events(self) -> Iterator[DagsterEvent]:
+        """Pops and yields all user-generated events that have been recorded from this context.
+
+        If consume_events has not yet been called, this will yield all logged events since the beginning of the op's computation. If consume_events has been called, it will yield all events since the last time consume_events was called. Designed for internal use. Users should never need to invoke this method.
+        """
+        events = self._events
+        self._events = []
+        yield from events
+
+    @public
+    def log_event(self, event: UserEvent) -> None:
+        """Log an AssetMaterialization, AssetObservation, or ExpectationResult from within the body of an op.
+
+        Events logged with this method will appear in the list of DagsterEvents, as well as the event log.
+
+        Args:
+            event (Union[AssetMaterialization, AssetObservation, ExpectationResult]): The event to log.
+
+        **Examples:**
+
+        .. code-block:: python
+
+            from dagster import op, AssetMaterialization
+
+            @op
+            def log_materialization(context):
+                context.log_event(AssetMaterialization("foo"))
+        """
+        if isinstance(event, AssetMaterialization):
+            self._events.append(
+                DagsterEvent.asset_materialization(self._step_execution_context, event)
+            )
+        elif isinstance(event, AssetObservation):
+            self._events.append(DagsterEvent.asset_observation(self._step_execution_context, event))
+        elif isinstance(event, ExpectationResult):
+            self._events.append(
+                DagsterEvent.step_expectation_result(self._step_execution_context, event)
+            )
+        else:
+            check.failed(f"Unexpected event {event}")
+
+    def add_output_metadata(
+        self,
+        metadata: Mapping[str, Any],
+        output_name: Optional[str] = None,
+        mapping_key: Optional[str] = None,
+    ) -> None:
+        """Add metadata to one of the outputs of an op.
+
+        This can only be used once per output in the body of an op. Using this method with the same output_name more than once within an op will result in an error.
```

Comment:
Btw I think this comment is out of date. I think sandy changed this so that you can call it multiple times and it merges/builds it as appropriate.

Reply:
checked the impl - this comment is still accurate